kernel/of/fdt.c

```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+
+    if (prop == NULL) {
+        // Linux default
+        return 2;
+    }
+
+    size_t num_cells = fdt_prop_num_cells(prop);
+    if (num_cells == 0) {
+        // This should never happen (Corrupted FDT)
+        // TODO: log warning
+        return (uint32_t)0;
+    }
+
+    if (num_cells > 1) {
+        // TODO: log a warning (this is unexpected)
+        //                     (not panic worthy though)
+    }
+
+    return fdt_prop_get_cell(prop, 0);
+}
+
+size_t
+fdt_prop_num_cells(struct fdt_prop* prop)
+{
+    return fdt_prop_val_len(prop) >> 2;
+}
+
+uint32_t
+fdt_prop_get_cell(struct fdt_prop* prop, size_t index)
+{
+    size_t num_cells = fdt_prop_num_cells(prop);
+    if (index >= num_cells) {
+        // Out of Bounds (Someone Goofed)
+        // TODO: log a warning
+        return 0;
+    }
+
+    uint32_t* val = fdt_prop_val(prop);
+    return be32toh(val[index]);
+}
+
+size_t
+fdt_node_num_register_blocks(struct fdt* fdt, struct fdt_node* node)
+{
+    struct fdt_prop* reg = fdt_get_prop_by_name(fdt, node, NULL, "reg");
+
+    if (reg == NULL) {
+        return (size_t)0;
+    }
+
+    uint32_t address_cells = fdt_node_get_address_cells(fdt, node);
+    uint32_t size_cells = fdt_node_get_address_cells(fdt, node);
+    uint32_t reg_block_cells = address_cells + size_cells;
+
+    if (reg_block_cells == 0) {
+        return 0;
+    }
+
+    // 4 bytes per cell
+    size_t reg_cell_len = fdt_prop_val_len(reg) >> 2;
+
+    return reg_cell_len / reg_block_cells;
+}
+
+void
+fdt_node_get_register_block(
+    struct fdt* fdt, struct fdt_node* node, size_t index, void** address, size_t* size
+)
+{
+    struct fdt_prop* reg = fdt_get_prop_by_name(fdt, node, NULL, "reg");
+
+    if (reg == NULL) {
+        goto err_exit;
+    }
+
+    uint32_t address_cells = fdt_node_get_address_cells(fdt, node);
+    uint32_t size_cells = fdt_node_get_address_cells(fdt, node);
+    uint32_t reg_block_cells = address_cells + size_cells;
+
+    void* raw_reg = fdt_prop_val(reg);
+    size_t raw_reg_len = fdt_prop_val_len(reg);
+
+    // How many bytes do we need "reg" to be so we can access index safely?
+    size_t min_size_for_index = (index * reg_block_cells * 4);
+
+    if (raw_reg_len < min_size_for_index) {
+        goto err_exit;
+    }
+
+    // Get a pointer to the start of the address in the property
+    uint32_t* address_ptr = ((uint32_t*)raw_reg) + (index * reg_block_cells);
+    // Size cells come after address cells
+    uint32_t* size_ptr = address_ptr + address_cells;
+
+    uint32_t msig_cell; // Most significant cell in 64-bit
+    uint32_t lsig_cell; // Least significant cell in 64-bit
+
+    switch (address_cells) {
+        case 1:
+            *address = (void*)(uintptr_t)be32toh(*address_ptr);
+            break;
+        case 2:
+            msig_cell = be32toh(address_ptr[0]);
+            lsig_cell = be32toh(address_ptr[1]);
+            *address = (void*)(((uint64_t)msig_cell << 32) | (uint64_t)lsig_cell);
```

Comment:
I'm sorry but I just see this and thing "too many parentheses it must be lisp"

I'm just trying to reduce parentheses

Reply:
They should be `uint32_t`'s though. And then combined to form one bigger `uint64_t`.

Reply:
Well combining two 32 bit ints into a 64 bit one is a common pattern, right?

Why not make a utils file for combining numbers. Like 4x 16 into a 64. Should be all inline.

Reply:
It's not that common, (this is literally just a `be64toh` written out) admittedly thinking about it I only wrote it like this because I was thinking about the restrictions of PCI ECAM space not the device tree. I'll swap it to just a `be64toh`